### PR TITLE
fix: Document Option.channel_types and properly support further channel types

### PR DESCRIPTION
**Pull request by NeloBlivion**
## Summary
- Finally documents `discord.Option.channel_types`
- Allows `input_type` to accept `discord.ForumChannel` and `discord.DMChannel`.
Ref. #1882 
- `channel_types` now *overrides* the `input_type`

## Information

- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [ ] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

**Comment by NeloBlivion** [comment]
So some clarification
1) `channel_types` only worked if you use `discord.SlashCommandOptionType.channel`, not standard channel classes (as pointed out in the linked issue)
2) For some reason, the Discord API actually allows private channels as an option type

Thus changes:
1) Now `channel_types` will **override** the `input_type`'s ChannelType
2) Both `discord.ForumChannel` and `discord.DMChannel` are now supported for `input_type`

File: discord/commands/options.py [hunk 1]
```diff
@@ -138,6 +148,10 @@ class Option:
         .. note::
 
             Does not validate the input value against the autocomplete results.
+    channel_types: Optional[List[:class:`discord.ChannelType`]]
+        A list of channel types that can be selected in this option.
+        Only applies to Options with an :attr:`input_type` of :class:`discord.abc.GuildChannel`.
```
Comment (plun1331) [inline discord/commands/options.py:153]: Maybe reference this as an `OptionType` instead

File: discord/commands/options.py [hunk 2]
```diff
@@ -138,6 +148,10 @@ class Option:
         .. note::
 
             Does not validate the input value against the autocomplete results.
+    channel_types: Optional[List[:class:`discord.ChannelType`]]
```
Comment (BobDotCom) [inline discord/commands/options.py:151]: We haven't manually gone in and updated old docstrings to the new annotation syntax, but for new entries we should definitely use it.
```suggestion
    channel_types: list[:class:`discord.ChannelType`] | None
```

Comment (Lulalaby) [reply discord/commands/options.py:151]: agreed

File: discord/commands/options.py (hunk 1 above)
Comment (NeloBlivion) [reply discord/commands/options.py:153]: Like this? 
```suggestion
        Only applies to Options with an :attr:`input_type` of :class:`discord.SlashCommandOptionType.channel`.
```

Comment (plun1331) [reply discord/commands/options.py:153]: Looks right

Comment (Lulalaby) [reply discord/commands/options.py:153]: Looks more left :p

Comment (plun1331) [reply discord/commands/options.py:153]: the line expanded to the right because more characters were added, therefore it is right not left

**Review by plun1331 — APPROVED** [review]

**Review by BobDotCom — APPROVED** [review]

**Pull request merged**